File: agent-mesh-py/python/agent_mesh/amesh/announce.py
"""``amesh announce`` — broadcast this agent's presence on the LAN.

Mirror of ``agent-mesh-cli/src/announce.rs``. Issues an ephemeral
AgentKey, stands up an mDNS announcer, and holds it alive either for
a bounded duration or until Ctrl-C.
"""

from __future__ import annotations

import asyncio
import sys
from pathlib import Path
from typing import List, Optional

from agent_mesh.core import AgentKey, AgentMetadata, UserKey
from agent_mesh.discovery import AnnounceConfig, Announcer

from . import util


async def run(
    home: Path,
    capabilities: List[str],
    role: str,
    host: Optional[str],
    duration: Optional[str],
) -> int:
    """Start the announcer and block until duration / Ctrl-C.

    ``duration=None`` runs forever (until SIGINT). Returns 0 on a
    clean shutdown, 1 if the user key is missing or duration parses
    badly.
    """
    key_path = home / "user.key"
    if not key_path.exists():
        print(
            f"error: load {key_path} — run `amesh keygen` first",
            file=sys.stderr,
        )
        return 1
    user = UserKey.load(str(key_path))

    resolved_host = host or util.current_hostname()
    meta = AgentMetadata(
        role=role,
        host=resolved_host,
        capabilities=list(capabilities),
        issued_at=util.now_rfc3339(),
    )
    # Ephemeral per-announce agent — never persisted, lives only as
    # long as the announcer task. Matches the Rust CLI's posture.
    agent = AgentKey.issue(user, meta)
    agent_fp = agent.fingerprint()
    user_fp = user.fingerprint()

    config = AnnounceConfig(
        agent_fp=agent_fp,
        user_fp=user_fp,
        capabilities=list(capabilities),
        role=role,
        host=resolved_host,
        port=0,
        # `amesh announce` is discovery-only — no transport bound, so
        # no agent pubkey is published. Peers who want to dial us
        # should use `amesh listen` instead.
        agent_pubkey=None,
    )
    handle = Announcer.start(config)

    print(
        f"announcing as agent_fp={agent_fp.hex()} user_fp={user_fp.hex()}"
    )
    print(
        f"  role={role} host={resolved_host} "
        f"capabilities={list(capabilities)!r}"
    )

    try:
        if duration is not None:
            try:
                dur_s = util.parse_duration(duration)
            except ValueError as e:
                handle.stop()
                print(f"error: {e}", file=sys.stderr)
                return 1
            print(f"  duration={dur_s}s")
            await asyncio.sleep(dur_s)
            print(f"announce duration {dur_s}s elapsed; stopping")
        else:
            print("  ctrl-c to stop")
            # Block forever; KeyboardInterrupt is handled in the
            # outer dispatcher.
            await asyncio.Event().wait()
    finally:
        handle.stop()
    return 0

File: agent-mesh-cli/src/announce.rs
//! `amesh announce` — broadcast this agent's presence on the LAN.
//!
//! Issues an ephemeral [`AgentKey`] for the announce session, then
//! starts an mDNS responder via [`agent_mesh_discovery::Announcer`].
//! Holds the announcer alive either for a bounded duration (when
//! `--duration` is supplied) or until the user hits Ctrl-C.

use crate::util;
use agent_mesh_discovery::{AnnounceConfig, Announcer};
use agent_mesh_protocol::{AgentKey, AgentMetadata, Caveats, UserKey};
use anyhow::{Context, Result};
use std::path::PathBuf;

/// Run the `announce` subcommand.
pub async fn run(
    home: PathBuf,
    capabilities: Vec<String>,
    role: String,
    host: Option<String>,
    duration: Option<String>,
) -> Result<()> {
    let key_path = home.join("user.key");
    let user = UserKey::load(&key_path)
        .with_context(|| format!("load {} — run `amesh keygen` first", key_path.display()))?;

    let host = host.unwrap_or_else(util::current_hostname);

    // Issue an ephemeral agent key for this announce session. The
    // mesh never sees the private half; the user/agent fingerprints
    // are what go on the wire in the mDNS TXT record.
    let agent = AgentKey::issue(
        &user,
        AgentMetadata {
            role: role.clone(),
            host: host.clone(),
            capabilities: capabilities.clone(),
            issued_at: util::now_rfc3339(),
            expires_at: None,
            caveats: Caveats::top(),
        },
    );

    let agent_fp = agent.fingerprint();
    let user_fp = user.fingerprint();

    let _handle = Announcer::start(AnnounceConfig {
        agent_fp,
        // `amesh announce` is discovery-only — no transport bound,
        // so we don't publish a pubkey. Peers who want to dial
        // should use `amesh listen` instead (Phase 2).
        agent_pubkey: None,
        user_fp,
        capabilities: capabilities.clone(),
        role: role.clone(),
        host: host.clone(),
        port: 0,
    })?;

    println!(
        "announcing as agent_fp={} user_fp={}",
        agent_fp.hex(),
        user_fp.hex()
    );
    println!("  role={role} host={host} capabilities={capabilities:?}");

    if let Some(d) = duration {
        let dur = util::parse_duration(&d)?;
        println!("  duration={dur:?}");
        tokio::time::sleep(dur).await;
        println!("announce duration {dur:?} elapsed; stopping");
    } else {
        println!("  ctrl-c to stop");
        tokio::signal::ctrl_c().await.context("wait for ctrl-c")?;
        println!("ctrl-c received; stopping");
    }
    Ok(())
}
